sgkit/stats/popgen.py

```diff
@@ -0,0 +1,47 @@
+from .aggregation import count_alleles
+
+from xarray import Dataset, DataArray
+from typing import Hashable, Iterable
+
+
+def diversity(
+    ds: Dataset,
+    allele_counts: Hashable = "variant_allele_counts",
+    windows: Iterable = None,
+    span_normalize: bool = True,
+) -> DataArray:
+    if allele_counts not in ds:
+        ds[allele_counts] = count_alleles(ds)
+    ac = ds[allele_counts]
+    an = ac.sum(axis=1)
+    n_pairs = an * (an - 1) / 2
+    n_same = (ac * (ac - 1) / 2).sum(axis=1)
+    n_diff = n_pairs - n_same
+    # Let's ignore missing data and division by zero for now.
+    pi = n_diff / n_pairs
+    # Because we're not providing any arguments on windowing, etc,
+    # we return the total over the whole region. Maybe this isn't
+    # the behaviour we want, but it's a starting point. Note that
+    # this is different to the tskit default behaviour where we
+    # normalise by the size of windows so that results
+    # in different windows are comparable. However, we don't have
+    # any information about the overall length of the sequence here
+    # so we can't normalise by it.
+    return pi.sum()
+
+
+def divergence(
```

Comment:
For consistency we should make the parameters ``ds1`` and ``ds2``, which have the same setup for ``allele_counts``. Either this, or we have one dataset and provide two selectors so specify the two subsets - but, then how would we cache the ``allele_counts`` computation? I think ``ds1``, ``ds2`` is the right approach.

What do you think @eric-czech ?

Reply:
`ds1`, `ds2` sounds good, though I'd have a slight preference for a signature like:

```python
def divergence(ds: Dataset, allele_counts: str, ...): ...
```

This would make the variable a required parameter for a single array and imply that the segregation must live formally as an axis in that array.  The advantages of this are:

- You don't have to deal with checking that `ac1`/`ac2` or that the variables in `ds1`/`ds2` have equal numbers of variants or alleles since that would be implicit in creating an array with both of them
- It would be more natural in a workflow that segregates more than two populations
- It would fit a little better in a pipeline:

```python
from sgkit import count_alleles
from sgkit.testing import simulate_genotype_call_dataset

ds = simulate_genotype_call_dataset(n_variant=100, n_sample=20, seed=0)
ds['sample_population'] = np.repeat([0, 1], ds.dims['samples'] // 2)
res = (
    ds
    .assign(
        allele_counts=ds
        .groupby('sample_population')
        .map(count_alleles)
        .rename_dims(dict(sample_population=populations))
    )
    .pipe(divergence, allele_counts='allele_counts')
)
```

The disadvantages are:

- The function would need to check that `ds.dims["populations"] == 2`, which is a little clunky 
- We should formalize the notion of a "populations" dimension if you go this route (i.e. we should use that same dimension name everywhere else)
- It assumes users are a bit more Xarray savvy 

I think finding common variables/dimension names for populations is inevitable so I'm sure that will come up again even if it's not here.  Otherwise, I don't feel strongly one way or the other but that little bit of extra generality could be worth it.

Reply:
Good ideas @eric-czech. The term "population" is problematic for lots of reasons, but I guess if we can come up with a term for "group of samples we want to aggregate together" I'm all for your approach.

I think at this point we should use xarray to its maximum potential/follow the idioms as closely as possible to establish the core functionality. Perhaps later then we can see how we can make another interface that builds on this that's simpler/easier for novice users, etc.

Reply:
>"group of samples we want to aggregate together"

`sample_group`, `sample_subset`, `subsample`?

Reply:
I like ``sample_group`` - seem unambiguous and flexible.

Reply:
For right now I did this:
`divergence(ds1: Dataset, ds2: Dataset,...):`

If/once we formalize a "groups" dimension, we can change it.

Reply:
OK, sounds good. We should probably create an issue for the ``group`` dimension, seems like a pretty fundamental aspect of the design.

Reply:
Issue opened: #224